### ISSUE TITLE
[Feature] Enable 3G Dongles

### ISSUE THREAD

**Issue by johnnyman727** [tessel/t2-cli#541]
To get 3G Dongles working, we'll need to update some dependencies on the device, add a udev rule, and finally add logic to the CLI to write the appropriate options to the `/etc/config/network` file. This issue is about that final piece which needs to be in t2-cli.

To see how to do this, check out https://wiki.openwrt.org/doc/recipes/3gdongle.

**Comment by johnnyman727** [comment]
This would still be a great feature and we can't use 3G dongles smoothly until it's added. If anyone is interested in contributing and needs more help, feel free to ping me here.

**Comment by shaochuancs** [comment]
Hi, @johnnyman727 , this would be a great feature. Looking forward that we can use 3G dongle automatically.

I opened [https://wiki.openwrt.org/doc/recipes/3gdongle](https://wiki.openwrt.org/doc/recipes/3gdongle) but feel completely confused (I'm a web developer and not familiar with OpenWRT/hardware/system programming). Does it mean we need to ssh to Tessel machine and run several commands there?

Thanks

**Comment by johnnyman727** [comment]
Hi @shaochuancs, thanks for getting involved!

> I opened https://wiki.openwrt.org/doc/recipes/3gdongle but feel completely confused (I'm a web developer and not familiar with OpenWRT/hardware/system programming)

No worries - I felt the same way when I first started perusing through OpenWRT docs (and still do sometimes). Feel free to ask me any questions that come up.

> Does it mean we need to ssh to Tessel machine and run several commands there?

Yes, that's exactly right. If your Tessel is connected to your wifi, you can get a shell with `t2 root`. That's the preferred way to work with Tessel's console (it uses `ssh` internally). If not, you can still get a shell over USB but you'll need to use a tool like `dterm` or `screen` (you'd use the command like `screen /dev/usbmodem1422`).

Typically, the way we build things for the CLI, is we'll use the shell to figure out what commands we need to run to get some feature to work. Once we have a list of commands to run, it becomes relatively simple to automate those commands with the CLI. 

As an example, [here is how we enable access points](https://github.com/tessel/t2-cli/blob/7d95332175d3a5d26f5f246d9166d913d1523eac/lib/tessel/access-point.js#L122-L143). You can see that we have a lot of commands that use:
```
tessel.simpleExec
```
This is basically a wrapper over:
```
tessel.connection.exec(SOME_COMMAND, () => {});
```
and it's a shorthand way of saying "run this shell command and then resolve a `Promise` once that command completes.

We have [one file](https://github.com/tessel/t2-cli/blob/7d95332175d3a5d26f5f246d9166d913d1523eac/lib/tessel/commands.js) where we list all the commands we use on Tessel (and pass to `simpleExec` or `tessel.connection.exec`). 

I hope that all makes sense - let me know if you have any questions!

**Comment by shaochuancs** [comment]
Hi, @johnnyman727, Thank you so much for the explanation! I'll read some OpenWRT docs and try to enable my 3G dongle.

Thanks

**Comment by johnnyman727** [comment]
No problem. Let me know how it goes!

**Comment by Jakehp** [comment]
+1

**Comment by opkiler22789** [comment]
+1

**Comment by brihijoshi** [comment]
Hey, @mittalshravika and I have been trying this issue. We used a shell over USB (screen) to update the "opkg" but are encountering the following issues (as in the screenshot)
![screen shot 2017-09-14 at 3 19 05 am](https://user-images.githubusercontent.com/22987501/30399971-170cf3d2-98f3-11e7-92f9-f4e02269cb56.png)

We tried looking up the error [here](https://communities.intel.com/thread/59889) and [here](https://forum.lede-project.org/t/opkg-update-error-wget-returned-4/637/7) but there has been no progress.

Could you please help?

@Frijol @johnnyman727

**Comment by johnnyman727** [comment]
Hi @brihijoshi, did you ensure that the Tessel has an internet connection before running `opkg update`?

**Comment by brihijoshi** [comment]
Hey, No it was not connected to the Internet. That is one reason why I opted for running the shell over USB. Thanks a lot!

**Comment by johnnyman727** [comment]
:+1: Good luck!

**Comment by brihijoshi** [comment]
Hey, so after we got our Tessel connected to the WiFi, we started downloading the packages as given in here - https://wiki.openwrt.org/doc/recipes/3gdongle

One issue came up in the command "opkg update" and an error is being thrown named "Signature Check Failed" as in the image below

![screen shot 2017-09-25 at 6 27 30 pm](https://user-images.githubusercontent.com/22987501/30807172-566e2f2e-a216-11e7-8845-891e56ff4b7a.png)

However, the remaining packages have been successfully installed. 

Any help would be amazing 😄

**Comment by Frijol** [comment]
possibly need to update your DNS address? https://forum.arduino.cc/index.php?topic=199167.0

**Comment by johnnyman727** [comment]
@brihijoshi Yes, I've seen that warning message shown when using `opkg update`. It's a problem with how something unrelated is configured but it shouldn't block OpenWRT from downloading system updates.

**Comment by bart** [comment]
Any updates so far? I also would like to use an USB 3G dongle from Huawei to access the internet with a Tessel 2 from all over the world.

**Comment by HipsterBrown** [comment]
@bart I've not had any progress, mostly because I don't have a USB 3G dongle for testing. If you would like to give it a shot, I'll provide any insight with OpenWRT and adding features to the CLI to help out.

**Comment by bart** [comment]
All right @HipsterBrown I'll see if I can do anything here. As far as I understood it's "just" about some linux settings that has to be made, right? I have a Huawei E173 which should work.

**Comment by bart** [comment]
All right it looks like OpenWRT is able to find my USB dongle without any problem. But when it comes to the chat script unfortunately it fails to connect. I'm using a Things Mobile SIM card here. No idea what the reason is.

```
Thu Dec 20 22:36:23 2018 local2.info chat[12186]: AT+CGDCONT=1,"IP","TM"^M^M
Thu Dec 20 22:36:23 2018 local2.info chat[12186]: +CME ERROR
Thu Dec 20 22:36:23 2018 local2.info chat[12186]:  -- failed
Thu Dec 20 22:36:23 2018 local2.info chat[12186]: Failed (ERROR)
Thu Dec 20 22:36:23 2018 daemon.debug pppd[12184]: Script USE_APN=TM DIALNUMBER=*99# /usr/sbin/chat -t5 -v -E -f /etc/chatscripts/3g.chat finished (pid 12185), status = 0x6
Thu Dec 20 22:36:23 2018 daemon.err pppd[12184]: Connect script failed
Thu Dec 20 22:36:24 2018 daemon.info pppd[12184]: Exit.
```

**Comment by HipsterBrown** [comment]
@bart Good to hear that OpenWRT is finding the USB dongle. To what chat script are you referring? Also, have you confirmed the SIM is activated?

**Comment by bart** [comment]
Ha! It works now. Was my bad because I put in a SIM card of wrong format... please, say nothing ;) All right, so now I must adapt the CLI to at least give users the possibility to configure their USB dongle that works out of the box without having to map it as serial device. Should I just check out the code myself or do you have something like a guide for adding new CLI commands?

**Comment by HipsterBrown** [comment]
@bart That's great you got it working! :clap: We don't have a guide for adding new command (good point though), but you can take a look at a previous PR like (https://github.com/tessel/t2-cli/pull/985/files) to see an example of adding a new command. Please let me know if you have any questions and feel free to open a PR at any point for early review.
